hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/MorphingPortableReader.java

```diff
@@ -47,7 +46,7 @@ public int readInt(String fieldName) throws IOException {
             case SHORT:
                 return super.readShort(fieldName);
             default:
-                throw new IncompatibleClassChangeError();
+                throw new IncompatibleClassChangeError("This type is incompatible to read from :" + fd.getType());
```

Comment:
I would pull out the IncompatibleClassChangeError into a single method instead of copy pasting it.

Also the message isn't that informative. Since the message doesn't include:
*  the class name
* the expected type. 

I would include the ClassDefinition which includes version? So the user has all information at hand to deal with the IncompatibleClassChangeError.